doc/authorization_policy_engine.md

```diff
@@ -0,0 +1,224 @@
+# Authorization policy engine
+
+The authorization decisions in SPIRE are determined by a policy engine which
+bases its decision on a rego policy and databindings with Open Policy Agent
+(OPA). 
+
+This is a sample configuration of the policy.
+
+```
+server {
+    auth_policy_engine {
+        opa_file_provider {
+            policy_path = "./conf/server/policy.rego"
+            permissions_path = "./conf/server/permissions.json"
+        }
+    }
+}
+```
+
+If the policy engine configuration is not set, it defaults to the [default SPIRE
+authorization policy](#default-configurations).
+
+# Details of the policy engine
+
+The policy engine is based on the [Open Policy Agent
+(OPA)](https://www.openpolicyagent.org/). This is configured via two
+components, the rego policy, and the permissions (or databindings in terms of
+OPA). 
+
+- The rego policy is a rego policy file defining how to authorize the API calls.
+- The permissions (or databindings) is a JSON blob that defines additional data
+  that can be used in the rego policy.
+
+In general there is an overlap in terms of which aspects of the policy can be
+part of the rego and databindings. However, the general rule is "How it is done"
+is part of the rego policy, and the "What does this apply to" is part of the
+permissions/databindings file.
+
+## Rego policy
+
+The rego policy defines how input to the policy engine is evaluated to produce the result used by SPIRE server for authorization decisions.
+
+This is defined by the result object:
+```
+result = {
+  "allow": true/false,
+  "allow_if_admin": true/false,
+  "allow_if_local": true/false,
+  "allow_if_downstream": true/false,
+  "allow_if_agent": true/false,
+}
+```
+
+The fields of the result are the following:
+- `allow`: a boolean that if true, will authorize the call
+- `allow_if_local`: a boolean that if true, will authorize the call only if the
+  caller is a local UNIX socket call
+- `allow_if_admin`: a boolean that if true, will authorize the call only if the
+  caller is a SPIFFE ID with the Admin flag set
+- `allow_if_downstream`: a boolean that if true, will authorize the call 
```

Comment:
I think we're missing `allow_if_agent` ?

Reply:
good catch - fixed